ADD IPN/webhook for stripe

ptibogxiv opened the pull request:
TODO: optimise code and uncustomize -ptibogxiv.net- but good and fonctionnal for test
TODO: apply and save fee
TODO: more langs for text

functionnal: 

- 3Dsecure: create invoice / paid / close
- customer add or delete
- payout update/ bank transfert

ptibogxiv committed:
ADD IPN/Webhook for stripe

TODO: optimise code and uncustomize of ptibogxiv.net but good and fonctionnal for test

ptibogxiv committed:
Update ipn.php

ptibogxiv commented:
not functionnal with last commit. i work on it

eldy commented on htdocs/public/stripe/ipn.php:
```diff
@@ -0,0 +1,335 @@
+$user->fetch(5);
+$user->getrights();
+
+if (! empty($conf->stripeconnect->enabled) && ! empty($conf->multicompany->enabled)) {
```
I have just pushed a PR so the stripe account can be saved into table llx_oauth_token 
Module OAuth must be on and MAIN_FEATURES_LEVEL set to 2 (development feature)
llx_oauth_token is the table dedicated to store token for oauth service like stripe connect and others, so you must get your data from this table instead of stripe_entity (that is a table that does not exists yet and should not exists since the table llx_oauth_token already exists:
This table contains a line like this one:

![image](https://user-images.githubusercontent.com/883887/37171729-1b0580de-230f-11e8-8f73-eb1b81cb5600.png)

Can you fix your code to use this existing table and forget stripe_entity ?

ptibogxiv replied:
@eldy i try to see on it tonight and tomorrow. For a connect standard account it's not a probleme because already user Oauth.

the probleme is for the custom or express account which only use API REST to get the account. 
we need to save the stripe account too for retrieve the stripe account balance or some additionnal operation
-> save them in llx_const ?

ptibogxiv replied:
Fix done but not realy functionnal/optimum if entity!=1 and not use stripe connect. Use and controle of stripe account is important  for a better security and stripe evolution of API

with new european rules in 2018/2019 PSD2, the use of stripe source for authenticate payment need the use of webhook for the feedback and  failing or dispute action

hregis replied:
@ptibogxiv 
what is the problem with multicompany ?

ptibogxiv replied:
@hregis no pb with your module 😉. Just stripe webhook request can't include this data in url and for action/payment we need this for use source payment (new European rules for authentificate payment or extra stripe payment gateway as sepa direct debit) and do correct automatic action in dolibarr linked with payment/payout/dispute/fee/refund. So an efficient and easy save of the couple stripe_account/entity is necessary whatever dolibarr only use stripe with your multicompany module (with same or different stripe account for each entity)  or enhance dolibarr+multicompany with my stripeconnect module who transforme dolibarr/stripe in marketplace (the first entity become the unique admin of stripe and can get some extra fee on each payment)

hregis replied:
@ptibogxiv 
you can not use "metadata" for send the entity ID ??

ptibogxiv replied:
@hregis no :'( stripe don't send it for all their callback ex: 3D secure, payout, failure
It won't be so simple :D
In fact we can, for some webhook callback, retrieve this metadata with some very complexe REST requests but it will be more loud (very loud) and slow that just save the trio stripe_account/entity/mode in the internal dolibarr's database. It's add a complementary security with bad request too.

hregis replied:
@ptibogxiv 
I do not understand, if it's callback is that you make a request? and if you make a request you can send metadata!? Stripe must return these metadata to you during callback no?

hregis replied:
@ptibogxiv 
each entity must be able to be autonomous and not depend on the first entity

ptibogxiv replied:
@hregis  
-yes but the callback for some don't return this directly. For stripe request we need the stripe_account (acc_xxxxxxxxxxxxx). All connect's callback has an stripe_account data in the header of the body response so it's more easy and direct to use it. 
ex payment with 3D secure
create 3Dsecure charge not charge but create a secondary card -> callback from stripe -> dolibarr create charge with this ->stripe return if it's ok.
-some automatic stripe's callback as payout have no metada in the response
-some API request in mode >1 entity and stripe not a principal account need include the stripe_account data to cal it.
- we must be compatible with standard an autonomous setting of stripe for entity>1. For the moment, webhook call isn't fully compatible for autonomous stripe entity and a standard account because standard stripe account hasn't stripe_account in the response and for some callback no metadata. 
=>The only solution for use webhook for entity >1 : link another standard stripe account to the principal with Oauth (with or without application fee) or use custom/express stripe account.

with this settings, we can let each entity autonomous or depend of the first entity according to business and/or technical settings of dolibarr.

ptibogxiv replied:
@hregis stripeconnect will support direct charge mode (each entity has each own customer and invoice directly) and separate transfert charge (principal is between customer and each company) for a standard/custom/express connect stripe's account. In dolibarr, all data/action keep autonomous and link to the good entity. It's important for me too.

ptibogxiv replied:
@hregis after reflexion and test, i found solution to solve pb for a standard and alone stripe account.
I will modify ipn.php in order to support both ipn?entity=X and without in connect mode
TODO: create url link  of ipn file to paste in stripe admin on admin/stripe.php

ptibogxiv committed:
NEW actions on card.php for company and member

ptibogxiv committed:
Update actions_stripe.class.php

ptibogxiv committed:
FIX dolibarr version

ptibogxiv committed:
FIX entity

TO DO: save stripe account in multicompany mode and without stripe connect in theadmin of stripe module because stripe doesn't send custom field for all request
RISK: confusion and error in payment and transfert bank without this

ptibogxiv committed:
FIX compatibility for multicompany (stripeconnect or not)

TODO: more test and create fee with stripe callback

ptibogxiv committed:
Update and fix

prepare integration for more callback support

ptibogxiv committed:
integration of fee's platform

ptibogxiv committed:
Update stripe.php

ptibogxiv committed:
Update stripe.php

ptibogxiv committed:
Update stripe.php

eldy committed:
Merge branch 'develop' into ptibogxiv-stripe
